### DISCUSSION TITLE
1.x Query API More Intuitive

### DISCUSSION THREAD

**Discussion by stealthco-tony** [sqlalchemy/sqlalchemy#11279] [Usage Questions]
Hi Team,

Know this might piss some people off, but can I ask what was the thought process behind switching to the new 2.x style querying? I know SQA will retain 1.x style {for some time, forever}, but just looking at the sytax on the website, 1.x seems significantly easier/clearer to read, which is a core Python coding tenant.

Review the following table from the [documentation](https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#migration-20-query-usage):
<img width="926" alt="Screenshot 2024-04-16 at 1 26 40 PM" src="https://github.com/sqlalchemy/sqlalchemy/assets/155121380/df6d464f-4264-4019-9ca3-a73b28c38255">

Just on the surface, 1.x style seems much more intuitive to read/write. Would love to learn what developers gain by migrating to 2.x style?

Thanks!

**Comment by gordthompson** [comment]
This section of the docs explains some of the reasoning behind the change.

https://docs.sqlalchemy.org/en/20/changelog/migration_14.html#orm-query-is-internally-unified-with-select-update-delete-2-0-style-execution-available

**Comment by CaselIT** [comment]
Hi,

The short story is to unify how select statements are constructed. It's vastly easier to have a single api for both core and orm.

Also query has some quirks that were impossible to fix without breaking user code in non-obvious ways

**Comment by zzzeek** [comment]
well the first issue was for the use of asyncio.    When you have a `Query` object, when do we need to await?  when there's IO.  When's that ?  Do we await on filter()?  (no)  do we await on all()?  (yes) do we await on first()?  (yes)   do we await on query.update()? (yes) but what if I want to run `query.update()` but also get the RETURNING rows, where do I do that?  (um.  I dunno, we can't add returning() ahead of it, awkward) What if I want `query.update()` with RETURNING and use it as a CTE in another statement?  How do we make it so we aren't awaiting that part?   (um.  I dunno)  What if I want a result object that I can stream asynchronously?  or get partitions (groups of rows) asynchronously?   Query would become a minefield of awaitables / non-awaitables, no clear pattern when IO is implied or not, and more composable patterns like being able to treat insert()/update()/delete() -> RETURNING the same as SELECT are mostly impossible without a much more awkward API.  solution.   **separate execution from query construction**.   `Session.execute()` is execution, operations on the inner `select()` are query construction.  A clean separation, which is clearer and easier to understand  (more verbose, but this is not 2005 or Perl).     

Second issue.  `Session.execute()` exists even before we had asyncio.   When do I use `Session.execute()` and when do I use `Query`?   how come I can send SELECT/INSERT/UPDATE/DELETE to `Session.execute()`?   Why do I need a `Query`?  Why is there `select().where()` but `query.filter()` ?  Why does `select().join()` build a weird useless subquery while `query().join()` adds a `.join()`?  But then I have a SELECT that I made with `union_all(select().union(select()), select())`, why can't I do that with query?  I tried `query.union_all(query.union(query)))` and it made a gigantic non-working mess.     

Also:

* [Why does my Query result not have the same number of rows as my COUNT](https://github.com/sqlalchemy/sqlalchemy/issues/4395)
* [Why does Query give me a tuple for multiple columns but I can't select just one column?](https://github.com/sqlalchemy/sqlalchemy/issues/5087)  
    * Maybe we should just start adding [arbitrary keyword arguments here and there without any composability or consistency like ](https://github.com/sqlalchemy/sqlalchemy/pull/5088/files) like `session.query().all(scalar=False)`  ?
* [Why can't I get ORM behaviors with insert()?](https://github.com/sqlalchemy/sqlalchemy/issues/7864)

> Just on the surface, 1.x style seems much more intuitive to read/write. 

it's more sparse and less typing for simple cases.  For more complex cases and many other areas of integration with other row-returning elements like INSERT/UPDATE/DELETE->RETURNING as well as with asyncio constructs, it falls over completely, APIs do not exist and expanding on the APIs that are there without a more composed structure would be a disaster.
